edb/schema/delta.py

```diff
@@ -2850,24 +2850,39 @@ def _cmd_tree_from_ast(
         field = parent_cls.get_field(propname)
 
         new_value: Any = astnode.value
+        old_value: Any = None
 
         if field.type is s_expr.Expression:
             if parent_cls.has_field(f'orig_{field.name}'):
                 orig_text = cls.get_orig_expr_text(
                     schema, parent_op.qlast, field.name)
             else:
                 orig_text = None
-            new_value = s_expr.Expression.from_ast(
-                astnode.value,
-                schema,
-                context.modaliases,
-                context.localnames,
-                orig_text=orig_text,
-            )
+
+            if astnode.value is not None:
+                new_value = s_expr.Expression.from_ast(
+                    astnode.value,
+                    schema,
+                    context.modaliases,
+                    context.localnames,
+                    orig_text=orig_text,
+                )
+            else:
+
+                # SetSpecialField 'expr' to None is `DROP EXPRESSION`,
+                # so we're going to need the old value in order to see
+                # that a change happened.
+                from . import pointers as s_pointers
+
+                pointer = cast(
+                    s_pointers.Pointer, schema.get(parent_op.classname))
```

Comment:
You cannot rely on `schema` at the `cmd_tree_from_ast` stage.  This has to go into the delta application phase, which, in this case would be `canonicalize_attributes` in `pointers.py`, and you'll likely also need to set up some sentinel in `_process_create_or_alter_ast` there too.

Reply:
I'm not sure I follow. The pointer in question *has* to already exist in the schema because we are in the `DROP EXPRESSION` branch. So there must be a pointer already defined and we can use it to get the `old_value`. What am I missing?

Reply:
> What am I missing?

The DDL script is not guaranteed to be executed statement-by-statement. Consider the following script:

```
ALTER TYPE Foo CREATE LINK foo := (...);
ALTER TYPE Foo ALTER LINK foo DROP EXPRESSION;
```

Also, whenever you find yourself importing a specific module into a general one, it's likely that you're not doing it in the correct spot.